### PR TITLE
driver: timer: power: npcx: Add system timer and power management support in npcx series.

### PR DESCRIPTION
**Pull request by MulinChao**
This PR is separated into three CLs.

This first CL introduces a kernel device driver implemented by the internal
64/32-bit timers in Nuvoton NPCX series. Via these two kinds of timer,
the driver provides a standard "system clock driver" interface. 

The second CL introduces power management driver that improves the efficiency
of ec operation by adjusting the chip’s power consumption to the level
of activity required by the application in npcx series.
And this CL implements one power state, PM_STATE_SUSPEND_TO_IDLE, with
two sub-states for Zephyr power management system.
Sub-state 0 - "Deep Sleep" mode with “Instant” wake-up if residency time
               is greater or equal to 1 ms
Sub-state 1 - "Deep Sleep" mode with "Standard" wake-up if residency time
               is greater or equal to 201 ms

The third CL introduces power management device support in uart device.

Athought the uart device clock will be stopped after ec entered sleep or
deep sleep state and restore its clock automatically (there is no need
to implement code for suspending and resuming devices manually.), 
we still need to check the current status of uart device when it 
wants to change its power state to LOW or SUSPEND power state. It is 
crucial to forbid ec enters sleep or deep sleep state when uart device is 
busy with transmitting data. Or we will observe broken characters on the 
uart console.

### CODE REVIEW

**Comment by MulinChao** [comment]
Digging the build error log further, I found pm_policy_low_power_devices() is not implemented in tests/kernel/profiling/profiling_api test suite. (This test suite enabled CONFIG_PM_POLICY_APP but has no default pm_policy_low_power_devices() function. I thought this function should be added in the test suite, not the pm driver itself.

File: drivers/timer/npcx_itim_timer.c
Comment (sjg20) [inline drivers/timer/npcx_itim_timer.c:32]: What are these last three bullet points about?

File: drivers/timer/npcx_itim_timer.c
Comment (sjg20) [inline drivers/timer/npcx_itim_timer.c:92]: drop outer ()

(in fact none is needed)

File: drivers/timer/npcx_itim_timer.c
Comment (sjg20) [inline drivers/timer/npcx_itim_timer.c:76]: minor point, but for naming I would suggest

cnt64h

and cnt64h_check

then use cnt64h below the loop. That makes it clearer what you are doing

File: drivers/timer/npcx_itim_timer.c
Comment (sjg20) [inline drivers/timer/npcx_itim_timer.c:104]: {} sadly to keep Zephyr happy

File: drivers/timer/npcx_itim_timer.c
Comment (sjg20) [inline drivers/timer/npcx_itim_timer.c:220]: Return

File: drivers/timer/npcx_itim_timer.c
Comment (sjg20) [inline drivers/timer/npcx_itim_timer.c:231]: Return

**Review by sjg20 — CHANGES_REQUESTED** [review]
again I prefer lower-case hex

File: drivers/timer/npcx_itim_timer.c
Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: Indeed, it looks ambiguous. I will modify them for better explanations.

Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: Done.

File: drivers/timer/npcx_itim_timer.c
Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: Done.

File: drivers/timer/npcx_itim_timer.c
Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: Indeed, that's clearer. Thanks.

Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: Done.

File: drivers/timer/npcx_itim_timer.c
Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c:106]: Done.  😄

File: drivers/timer/npcx_itim_timer.c
Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: Done.

File: drivers/timer/npcx_itim_timer.c
Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: Done.

**Comment by MulinChao** [comment]
> again I prefer lower-case hex

Done.

**Review by sjg20 — APPROVED** [review]

File: drivers/clock_control/clock_control_npcx.c
Comment (yperess) [inline drivers/clock_control/clock_control_npcx.c:132]: Could you add a closing comment to make sure that the if/endif are associated?

```
#endif /* CONFIG_PM */
```

File: drivers/espi/host_subs_npcx.c
Comment (yperess) [inline drivers/espi/host_subs_npcx.c:1063]: This is more readable as:

```
if (IS_ENABLED(CONFIG_PM)) {
  ...
}
```

File: drivers/serial/uart_npcx.c
Comment (yperess) [inline drivers/serial/uart_npcx.c:378]: Same here, prefer

```
if (IS_ENABLED(CONFIG_UART_INTERRUPT_DRIVEN))
        return ...
else
        return 0;
```

File: drivers/serial/uart_npcx.c
Comment (yperess) [inline drivers/serial/uart_npcx.c:432]: Please add a closing comment:

```
#endif /* CONFIG_PM_DEVICE */
```

File: drivers/timer/npcx_itim_timer.c
Comment (yperess) [inline drivers/timer/npcx_itim_timer.c:95]: Here too, prefer `IS_ENABLED` for readability.

File: drivers/timer/npcx_itim_timer.c
Comment (yperess) [inline drivers/timer/npcx_itim_timer.c:194]: Closing comment here too please.

File: drivers/timer/npcx_itim_timer.c
Comment (yperess) [inline drivers/timer/npcx_itim_timer.c:319]: Closing comments on these too please.

File: soc/arm/nuvoton_npcx/npcx7/power.c
Comment (yperess) [inline soc/arm/nuvoton_npcx/npcx7/power.c:69]: If we just have these as static will they not be taken out by the optimizer anyway if `!IS_ENABLED(CONFIG_SOC_POWER_MANAGEMENT_TRACE)`? Putting these in will allow us to use the more readable `IS_ENABLED` further down.

File: soc/arm/nuvoton_npcx/npcx7/power.c
Comment (yperess) [inline soc/arm/nuvoton_npcx/npcx7/power.c:115]: Referencing the comment above, if we always define these we can just use `if (IS_ENABLED(...` here.

File: soc/arm/nuvoton_npcx/npcx7/power.c
Comment (yperess) [inline soc/arm/nuvoton_npcx/npcx7/power.c:121]: Same here

File: soc/arm/nuvoton_npcx/npcx7/power.c
Comment (yperess) [inline soc/arm/nuvoton_npcx/npcx7/power.c:156]: Same here

**Review by yperess — CHANGES_REQUESTED** [review]
Overall it looks good, minor nits here and there.

File: soc/arm/nuvoton_npcx/common/reg/reg_def.h
Comment (keith-zephyr) [inline soc/arm/nuvoton_npcx/common/reg/reg_def.h:1149]: The NPCX7nxC datasheet shows all register names as ITPRE32, ITCTS32, and ITCNT32.  I'm not sure about other datasheets.  But I think it improves readability of the driver if the 32-bit timer always includes "32" on the register name and same with the 64-bit timer (add "64").

File: soc/arm/nuvoton_npcx/common/reg/reg_def.h
Comment (keith-zephyr) [inline soc/arm/nuvoton_npcx/common/reg/reg_def.h:1161]: Add "64" to the register names.

File: drivers/timer/npcx_itim_timer.c
Comment (keith-zephyr) [inline drivers/timer/npcx_itim_timer.c:121]: Can you put a upper bound on waiting of ITEN to set/clear for all instances?

File: drivers/timer/npcx_itim_timer.c
Comment (keith-zephyr) [inline drivers/timer/npcx_itim_timer.c:281]: The datasheet indicates to wait 92us after changing CLKSEL before enabling the module.  Probably a good idea to wait inline after updates to ITCTS.

File: drivers/timer/npcx_itim_timer.c
Comment (keith-zephyr) [inline drivers/timer/npcx_itim_timer.c:155]: This could be guarded with IS_ENABLED(CONFIG_TICKLESS_KERNEL) since delta_ticks is unused otherwise.

File: soc/arm/nuvoton_npcx/npcx7/power.c
Comment (keith-zephyr) [inline soc/arm/nuvoton_npcx/npcx7/power.c:112]: If you change the parameters to npcx_power_enter_system_sleep() to enums, then this code becomes a lot more readable.  For example:
    npcx_power_enter_system_sleep(DEEP_SLEEP, INSTANT_WAKE_UP);
    npcx_power_enter_system_sleep(DEEP_SLEEP, STANDARD_WAKE_UP);

File: drivers/serial/uart_npcx.c
Comment (keith-zephyr) [inline drivers/serial/uart_npcx.c:376]: Typo: "No need for polling mode"

File: drivers/serial/uart_npcx.c
Comment (keith-zephyr) [inline drivers/serial/uart_npcx.c:403]: Please put an upper bound on this loop.

File: drivers/clock_control/clock_control_npcx.c
Comment (MulinChao) [reply drivers/clock_control/clock_control_npcx.c]: Done.

File: drivers/espi/host_subs_npcx.c
Comment (MulinChao) [reply drivers/espi/host_subs_npcx.c]: Done.

File: drivers/serial/uart_npcx.c
Comment (MulinChao) [reply drivers/serial/uart_npcx.c]: Done.

File: drivers/serial/uart_npcx.c
Comment (MulinChao) [reply drivers/serial/uart_npcx.c]: Done.

File: drivers/timer/npcx_itim_timer.c
Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c:96]: Replacing this with IS_ENABLED() macro will cause a compiler error since 'cyc_sys_compensated' is undeclared if CONFIG_PM is not enabled. If I remove line 69 & 74, we need to move 'cyc_evt_enter_deep_idle' variable to line 309 in case 'unused variable' warning. Does it improve readability if I do so?

File: drivers/timer/npcx_itim_timer.c
Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: Done.

File: drivers/timer/npcx_itim_timer.c
Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: Done.

File: soc/arm/nuvoton_npcx/npcx7/power.c
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/npcx7/power.c]: Good idea. I will replace them with IS_ENABLED() macro later.

Comment (MulinChao) [reply soc/arm/nuvoton_npcx/npcx7/power.c]: Done.

File: soc/arm/nuvoton_npcx/npcx7/power.c
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/npcx7/power.c]: Done.

File: soc/arm/nuvoton_npcx/npcx7/power.c
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/npcx7/power.c]: Done.

File: soc/arm/nuvoton_npcx/npcx7/power.c
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/npcx7/power.c]: Done.

File: soc/arm/nuvoton_npcx/common/reg/reg_def.h
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/common/reg/reg_def.h]: The original thought is to keep the ITCNT register and ITCNT_XXX field as the same prefix no matter 32-bit or 64-bit timers. But from the improving readability point of view, I agree with you. I will change all register's names to meet our datasheet and modified the prefix of the field name from NPCX_ITCTS_ to NPCX_ITCTSXX_.

Comment (MulinChao) [reply soc/arm/nuvoton_npcx/common/reg/reg_def.h]: Done.

File: soc/arm/nuvoton_npcx/common/reg/reg_def.h
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/common/reg/reg_def.h]: Done.

File: drivers/timer/npcx_itim_timer.c
Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: Indeed. I will add a delay later.

Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: Done.

File: drivers/timer/npcx_itim_timer.c
Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: Done.

File: soc/arm/nuvoton_npcx/npcx7/power.c
Comment (MulinChao) [reply soc/arm/nuvoton_npcx/npcx7/power.c]: Done.

File: drivers/serial/uart_npcx.c
Comment (MulinChao) [reply drivers/serial/uart_npcx.c]: Done.

File: drivers/serial/uart_npcx.c
Comment (MulinChao) [reply drivers/serial/uart_npcx.c]: Done. The same assumption as above.

**Comment by MulinChao** [comment]
Hi @yperess and @keith-zephyr :
I have finished most of the requests and left few open items for further discussion. Thanks for the effort of reviewing this.

File: drivers/timer/npcx_itim_timer.c
Comment (yperess) [reply drivers/timer/npcx_itim_timer.c:96]: I see, up to you, personally I prefer to set line 71 to:
```
__maybe_unused static uint64_t cyc_sys_compensated;
```

That avoids the warning. Though looking at the code further I think you can just always declare cyc_sys_compensated which will default to 0, then you don't even need this if statement here and you can just always `+ cyc_sys_compensated`. The compiler should have an easy enough time figuring out that `cyc_sys_compensated` is always 0 and replace it accordingly.

* I recognize that I'm always in favor of offloading work to the compiler/optimizer in favor of simpler code, if you disagree I'm not going to block on this, just throwing my 2 cents.

File: drivers/timer/npcx_itim_timer.c
Comment (keith-zephyr) [inline drivers/timer/npcx_itim_timer.c:122]: I would still like an upper bound put on the while loops.

File: drivers/timer/npcx_itim_timer.c
Comment (keith-zephyr) [inline drivers/timer/npcx_itim_timer.c:160]: Same here - please prevent an infinite loop in case of a hardware failure.

**Review by keith-zephyr — APPROVED** [review]

File: drivers/timer/npcx_itim_timer.c
Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c:121]: Sure. I will submit the other PR for this topic later.

File: drivers/timer/npcx_itim_timer.c
Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: _maybe_unused seems doesn't work on Zephyr. But I use "__unused" instead of it. I agree with you that " offloading work to the compiler/optimizer in favor of simpler code" improves readability. Thanks for your comment. I learn a lot.

Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: Done.

File: drivers/timer/npcx_itim_timer.c
Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: Hi @keith-zephyr 
I abstract all infinite loops and put them in two inline functions, npcx_itim_evt_enable() and npcx_itim_evt_disable().
After discussing with my hardware colleague, we think it is hardware proven behavior. (ITEN will be high after one clock from writing it as 1. Clearing it will take effect immediately.) Since it only costs 30us at most, do you think we need a timeout mechanism? If so, I will add the other PR for this later.

File: drivers/timer/npcx_itim_timer.c
Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: Done. Please see below. Thanks.

File: drivers/timer/npcx_itim_timer.c
Comment (keith-zephyr) [reply drivers/timer/npcx_itim_timer.c]: These particular timeouts would only occur if there is a failure in the silicon (which I agree is very unlikely).  But it's good to eliminate unbounded loops whenever possible.  Please add this with another PR, but it's not high priority.

Comment (MulinChao) [reply drivers/timer/npcx_itim_timer.c]: Sure. I will add the other PR for unbounded loops after this PR is merged.

**Comment by MulinChao** [comment]
> Digging the build error log further, I found pm_policy_low_power_devices() is not implemented in tests/kernel/profiling/profiling_api test suite. (This test suite enabled CONFIG_PM_POLICY_APP but has no default pm_policy_low_power_devices() function. I thought this function should be added in the test suite, not the pm driver itself.

A new [CL](https://github.com/zephyrproject-rtos/zephyr/pull/31697/commits/0358aeed8aa65160e1c6c8367160a1c2d62f2a7e)  is added in this PR for the build issue of tests/kernel/profiling/profiling_api test suite in case buildkite/zephyr blocks this PR.

**Review by keith-zephyr — APPROVED** [review]

**Comment by MulinChao** [comment]
Rebasing since [PR31578](https://github.com/zephyrproject-rtos/zephyr/pull/31578)

**Pull request merged**
